File: UniversalBre/atom_status.cpp
#include "atom_status.h"
#include "utility.h"

core::atom_status::atom_status(const wchar_t atom) : atom_(atom)
{
    is_digit_ = utility::is_digit(atom);
    is_whitespace_ = utility::is_whitespace(atom);
    is_alpha_ = utility::is_alpha(atom);
    is_alpha_digit_ = utility::is_alpha_digit(atom);
    is_alpha_digit_underscore_ = utility::is_alpha_digit(atom) || is_underscore();
    is_break_char_ = utility::is_break_char(atom);
    is_integer_break_char_ = utility::is_integer_break_char(atom);
}

bool core::atom_status::is_dot() const
{
    return atom_ == '.';
}

bool core::atom_status::is_digit() const
{
    return is_digit_;
}

bool core::atom_status::is_whitespace() const
{
    return is_whitespace_;
}

bool core::atom_status::is_break_char() const
{
    return is_break_char_;
}

bool core::atom_status::is_integer_break_char() const
{
    return is_integer_break_char_;
}

bool core::atom_status::is_alpha() const
{
    return is_alpha_;
}

bool core::atom_status::is_alpha_digit() const
{
    return is_alpha_digit_;
}

bool core::atom_status::is_alpha_digit_underscore() const
{
    return is_alpha_digit_underscore_;
}

bool core::atom_status::is_empty() const
{
    return atom_ == L'\x0';
}

bool core::atom_status::is_empty_or_whitespace() const
{
    return is_empty() || is_whitespace();
}

bool core::atom_status::breaks_any() const
{
    return is_empty() || is_whitespace() || is_break_char();
}

bool core::atom_status::breaks_any_integer() const
{
    return is_empty() || is_whitespace() || is_integer_break_char();
}

bool core::atom_status::breaks_any_string() const
{
    return atom_ == '"';
}

bool core::atom_status::is_underscore() const
{
    return atom_ == '_';
}

bool core::atom_status::is_identifier_char() const
{
    return is_alpha_digit_underscore();
}

wchar_t core::atom_status::get_atom() const
{
    return atom_;
}

File: src/atom_status.rs
#[derive(Debug, Clone)]
pub struct AtomStatus {
    atom: char,
}

impl AtomStatus {
    pub fn init(atom: char) -> AtomStatus {
        AtomStatus { atom: atom }
    }

    pub fn is_dot(&self) -> bool {
        self.atom == '.'
    }

    pub fn is_digit(&self) -> bool {
        self.atom.is_digit(10)
    }

    pub fn is_whitespace(&self) -> bool {
        let whitespace_chars: Vec<char> = vec![' ', '\r', '\n', '\t'];
        whitespace_chars.contains(&self.atom)
    }

    pub fn is_newline_char(&self) -> bool {
        let newline_chars: Vec<char> = vec!['\r', '\n'];
        newline_chars.contains(&self.atom)
    }

    pub fn is_break_char(&self) -> bool {
        let breaking_chars: Vec<char> = vec![
            '(', ')', '+', '-', '*', '/', '~', '.', '<', '>', '=', '&', '|', ';', '{', '}', ',',
            ']', '[', ':',
        ];
        breaking_chars.contains(&self.atom)
    }

    pub fn is_integer_break_char(&self) -> bool {
        let integer_breaking_chars: Vec<char> = vec![
            '(', ')', '+', '-', '*', '/', '~', '<', '>', '=', '&', '|', ';', '{', '}', ',', ']',
            '[', ':',
        ];
        integer_breaking_chars.contains(&self.atom)
    }

    pub fn is_alpha(&self) -> bool {
        self.atom.is_alphabetic()
    }

    pub fn is_alpha_digit(&self) -> bool {
        self.is_alpha() || self.is_digit()
    }

    pub fn is_alpha_digit_underscore(&self) -> bool {
        self.is_alpha_digit() || self.is_underscore()
    }

    pub fn is_empty(&self) -> bool {
        self.atom == '\0'
    }

    pub fn is_empty_or_whitespace(&self) -> bool {
        self.is_empty() || self.is_whitespace()
    }

    pub fn breaks_any(&self) -> bool {
        self.is_empty_or_whitespace() || self.is_break_char()
    }

    pub fn breaks_any_integer(&self) -> bool {
        self.is_empty_or_whitespace() || self.is_integer_break_char()
    }

    pub fn breaks_any_string(&self) -> bool {
        self.atom == '"'
    }

    pub fn is_underscore(&self) -> bool {
        self.atom == '_'
    }

    pub fn is_identifier_char(&self) -> bool {
        self.is_alpha_digit_underscore()
    }

    pub fn get_atom(&self) -> char {
        self.atom
    }
}
